contrib/babelfishpg_common/src/datetime.c

```diff
@@ -797,4 +804,532 @@ datetime_to_numeric(PG_FUNCTION_ARGS)
+				break;
+			case DTK_MONTH:
+				if(dttype == TIME) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type time",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, num, 0, 0, 0, 0, 0);
+				break;
+			case DTK_WEEK:
+				if(dttype == TIME) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type time",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, num, 0, 0, 0, 0);
+				break;
+			case DTK_DAY:
+			case DTK_DOY:
+				if(dttype == TIME) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type time",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, num, 0, 0, 0);
+				break;
+			case DTK_HOUR:
+				if(dttype == DATE) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type date",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, num, 0, 0);
+				break;
+			case DTK_MINUTE:
+				if(dttype == DATE) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type date",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, num, 0);
+				break;
+			case DTK_SECOND:
+				if(dttype == DATE) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type date",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum(num));
+				break;
+			case DTK_MILLISEC:
+				if(dttype == DATE) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type date",
+				 	lowunits);
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum((float) num * 0.001));
+				break;
+			case DTK_MICROSEC:
+				if(dttype == SMALLDATETIME || dttype == DATETIME || dttype == DATE) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type %s",
+				 	lowunits, datetypeName(dttype));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum((float) num * 0.000001));
+				break;
+			case DTK_NANO:
+				if(dttype == SMALLDATETIME || dttype == DATETIME || dttype == DATE) {
+					elog(ERROR, "the datepart %s is not supported by date function dateadd for data type %s",
+				 	lowunits, datetypeName(dttype));
+				}
+				interval = (Interval *) DirectFunctionCall7(make_interval, 0, 0, 0, 0, 0, 0, Float8GetDatum((float) num * 0.000000001));
+				break;
+			default:
+				elog(ERROR, "the datepart %s is not supported by date function dateadd for data type %s",
+				 	lowunits, datetypeName(dttype));
+				break;
+		}
+	} else {
+		elog(ERROR, "the datepart %s is not supported by date function dateadd for data type %s",
```

Comment:
And can we make the errmsg in the same code appear place , not repeat so many times